docs/madr/003-arns-undername-limits.md

```diff
@@ -0,0 +1,237 @@
+    ANTContract->>AR.IO Gateway: Return records with priority data
+    AR.IO Gateway->>ARIOContract: Get undername limit
+    ARIOContract->>AR.IO Gateway: Return undername limit
+    note over AR.IO Gateway: Sort records by ANT priority, fallback to alphabetical
+    note over AR.IO Gateway: Validate undernames against undername limit
+    alt Undername limit exceeded
+        AR.IO Gateway->>Client: Return 402 (Payment required)
+    else Undername limit not exceeded
+        AR.IO Gateway->>Client: Resolve undername
+    end
+```
+
+## Decision Drivers
+
+- Honor ANT priority ordering when available
+- Consistent enforcement of undername limits across AR.IO gateways
+- Simple fallback mechanism if ANT does not return or contain priority data
+- Respects existing ANT records object type (minimal changes to [ar-io-node] & [ar-io-sdk])
+
+## Considered Options
+
+### Option 1: ANTs provide sorted array of records
+
+Lua does not guarantee the order of keys in a table. This raises some issues with the existing `getRecords` API that returns a table of records on an ANT. To circumvent this, we could modify the return type of `getRecords` or introduce a new API to a sorted array of records, based on the ANT priority. This would require a change to the [ar-io-sdk] and [ar-io-node] to leverage this updated API.
+
+ANT process with updated logic:
+
+```lua
+Records = (
+  '@': {
+    transactionId: '0x123',
+    ttlSeconds: 900
+  }
+)
+function getRecords(name: string)
+  return {
+    -- order of keys is not guaranteed in lua table
+    ["@"] = { transactionId = "0x123", ttlSeconds = 1000000, priority = 0 },
+    ["undername1"] = { transactionId = "0x123", ttlSeconds = 1000000, priority = 1 },
+    ["undername2"] = { transactionId = "0x123", ttlSeconds = 1000000, priority = 2 }
+  }
+end
+
+--- new logic to sort records based on some sort attributes
+function getSortedRecords(name: string):
+  local sortedRecords = {}
+  -- ANT decides how to sort records
+  for key, value in pairs(Records) do
+    table.insert(sortedRecords, {
+      record = key,
+      transactionId = value.transactionId,
+      ttlSeconds = value.ttlSeconds,
+    })
+  end
+  -- apply sort based on some attribute of the ANT (e.g. alphabetical)
+  table.sort(sortedRecords, function(a, b)
+    -- '@' record should always be first
+    if a.record == "@" then return true end
+    if b.record == "@" then return false end
+    return a.record > b.record
+  end)
+  return sortedRecords
+end
+```
+
+This array of sorted records would then be what the [ar-io-node] receives and enforces against the undername limit.
+
+### Option 2: ANTS provide priority data in records
+
+ANTs store additional information in the state for each record, indicating the priority a name. The [ar-io-node] would respect this priority when resolving undernames. If the ANT does not return priority attribute for a record, the [ar-io-node] would sort undernames alphabetically.
```

Comment:
`sort undernames alphabetically`
I think that's fine in theory. But just be mindful that this means different things in different use cases. e.g. lexical vs. numeric sorting of numbers.

Reply:
specified they will default lexical sorting

Reply:
[90be780](https://github.com/ar-io/ar-io-node/pull/272/commits/90be780969e021f3f9cfee168c331a143129ec1e)